beacon-chain/cache/registration.go

```diff
@@ -0,0 +1,67 @@
+package cache
+
+import (
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+)
+
+// RegistrationCache is used to store the cached results of an Validator Registration request.
+// beacon api /eth/v1/validator/register_validator
+type RegistrationCache struct {
+	sync.RWMutex
+	indexToRegistration map[primitives.ValidatorIndex]*ethpb.ValidatorRegistrationV1
+}
+
+// NewRegistrationCache initializes the map and underlying cache.
+func NewRegistrationCache() *RegistrationCache {
+	return &RegistrationCache{
+		indexToRegistration: make(map[primitives.ValidatorIndex]*ethpb.ValidatorRegistrationV1),
+	}
+}
+
+// GetRegistrationByIndex returns the registration by index in the cache and also removes items in the cache if expired.
+func (regCache *RegistrationCache) GetRegistrationByIndex(id primitives.ValidatorIndex) (*ethpb.ValidatorRegistrationV1, error) {
+	regCache.RLock()
+	v, ok := regCache.indexToRegistration[id]
+	if !ok {
+		defer regCache.RUnlock()
+		return nil, errors.Wrapf(ErrNotFoundRegistration, "validator id %d", id)
+	}
+	if RegistrationTimeStampExpired(v.Timestamp) {
+		regCache.RUnlock()
+		regCache.Lock()
+		defer regCache.Unlock()
+		delete(regCache.indexToRegistration, id)
+		return nil, errors.Wrapf(ErrNotFoundRegistration, "validator id %d", id)
+	}
+	defer regCache.RUnlock()
+	return v, nil
+}
```

Comment:
```suggestion
func (regCache *RegistrationCache) GetRegistrationByIndex(id primitives.ValidatorIndex) (*ethpb.ValidatorRegistrationV1, error) {
	regCache.RLock()
	v, ok := regCache.indexToRegistration[id]
	if !ok {
		regCache.RUnlock()
		return nil, errors.Wrapf(ErrNotFoundRegistration, "validator id %d", id)
	}
	if RegistrationTimeStampExpired(v.Timestamp) {
		regCache.RUnlock()
		regCache.Lock()
		defer regCache.Unlock()
		delete(regCache.indexToRegistration, id)
		return nil, errors.Wrapf(ErrNotFoundRegistration, "validator id %d", id)
	}
	regCache.RUnlock()
	return v, nil
}
```
For these RUnlock calls, you dont need a defer really because you've already read the data you need and you are returning on the next line for both of these.